File: iAsync.ObjC/lib/JFFStoreKit/JFFStoreKit/asyncAdapters/asyncSKProductRequest.m
#import "asyncSKProductRequest.h"

#import "JFFStoreKitCanNoLoadProductError.h"
#import "JFFStoreKitInvalidProductIdentifierError.h"

@interface JFFAsyncSKProductsRequestAdapter : NSObject <
JFFAsyncOperationInterface,
SKProductsRequestDelegate
>

@end

@implementation JFFAsyncSKProductsRequestAdapter
{
    SKProductsRequest                *_request;
    JFFDidFinishAsyncOperationCallback _finishCallback;
    SKProductsResponse               *_response;
    NSString                         *_productIdentifier;
}

- (instancetype)initWithProductIdentifier:(NSString *)productIdentifier
{
    self = [super init];
    
    if (self) {
        _productIdentifier = productIdentifier;
        SKProductsRequest *request= [[SKProductsRequest alloc] initWithProductIdentifiers:
                                     [NSSet setWithObject:productIdentifier]];
        
        _request   = request;
        request.delegate = self;
    }
    
    return self;
}

- (void)asyncOperationWithResultCallback:(JFFDidFinishAsyncOperationCallback)finishCallback
                         handlerCallback:(JFFAsyncOperationChangeStateCallback)handlerCallback
                        progressCallback:(JFFAsyncOperationProgressCallback)progressCallback
{
    _finishCallback = [finishCallback copy];
    
    [_request start];
}

- (void)doTask:(JFFAsyncOperationHandlerTask)task
{
    NSCParameterAssert(task <= JFFAsyncOperationHandlerTaskCancel);
    
    if (task == JFFAsyncOperationHandlerTaskCancel) {
        [_request cancel];
    }
}

#pragma mark SKRequestDelegate

- (void)requestDidFinish:(SKRequest *)request
{
    NSArray *products = _response.products;
    if ([products count] > 0) {
        
        SKProduct *product = [products firstMatch:^BOOL(SKProduct *product) {
            return [product.productIdentifier isEqualToString:_productIdentifier];
        }];
        
        if (!product) {
            
            NSString *log = [[NSString alloc] initWithFormat:@"requestDidFinish products does not contains product with id: %@", _productIdentifier];
            [[JLogger sharedJLogger] logError:log];
            product = [products lastObject];
        }
        
        _finishCallback(product, nil);
    } else {
        
        NSString *invalidIdentifier = [_response.invalidProductIdentifiers lastObject];
        
        JFFStoreKitCanNoLoadProductError *error = ([invalidIdentifier isEqualToString:_productIdentifier])
        ?[JFFStoreKitInvalidProductIdentifierError new]
        :[JFFStoreKitCanNoLoadProductError new];
        
        error.productIdentifier = _productIdentifier;
        _finishCallback(nil, error);
    }
}

- (void)request:(SKRequest *)request didFailWithError:(NSError *)error
{
    error = error?:[[JFFSilentError alloc] initWithDescription:@"SKRequest no inet connection"];
    _finishCallback(nil, error);
}

#pragma mark SKProductsRequestDelegate

- (void)productsRequest:(SKProductsRequest *)request didReceiveResponse:(SKProductsResponse *)response
{
    _response = response;
}

@end

JFFAsyncOperation skProductLoaderWithProductIdentifier(NSString *identifier)
{
    JFFAsyncOperationInstanceBuilder factory = ^id< JFFAsyncOperationInterface >() {
        return [[JFFAsyncSKProductsRequestAdapter alloc] initWithProductIdentifier:identifier];
    };
    return buildAsyncOperationWithAdapterFactory(factory);
}

File: iAsync.Swift/lib/JStoreKit/JStoreKit/asyncAdapters/asyncSKProductRequest.swift
//
//  asyncSKProductRequest.swift
//  JStoreKit
//
//  Created by Vladimir Gorbenko on 29.09.14.
//  Copyright (c) 2014 EmbeddedSources. All rights reserved.
//

import Foundation

import JAsync
import JUtils
import StoreKit

private class JAsyncSKProductsRequestAdapter : NSObject, JAsyncInterface, SKProductsRequestDelegate {

    typealias ResultType = SKProduct
    
    private let request: SKProductsRequest
    private var finishCallback: JAsyncTypes<ResultType>.JDidFinishAsyncCallback?
    private var response: SKProductsResponse?
    private let productIdentifier: String
    
    init(productIdentifier: String) {
        
        self.productIdentifier = productIdentifier
        self.request           = SKProductsRequest(productIdentifiers:Set([productIdentifier]))
        
        super.init()
        
        request.delegate = self
    }
    
    func asyncWithResultCallback(
        finishCallback: JAsyncTypes<ResultType>.JDidFinishAsyncCallback,
        stateCallback: JAsyncChangeStateCallback,
        progressCallback: JAsyncProgressCallback)
    {
        self.finishCallback = finishCallback
        
        request.start()
    }
    
    func doTask(task: JAsyncHandlerTask) {
        
        assert(task.rawValue <= JAsyncHandlerTask.Cancel.rawValue)
    
        if task == .Cancel {
            request.cancel()
        }
    }
    
    var isForeignThreadResultCallback: Bool {
        return false
    }
    
    func requestDidFinish(request: SKRequest!) {
        
        let products = response!.products as! [SKProduct]
        if products.count > 0 {
        
            var product = firstMatch(products) { (product: SKProduct) -> Bool in
                return product.productIdentifier == self.productIdentifier
            }
            
            if product == nil {
                
                let log = "requestDidFinish products does not contains product with id: \(productIdentifier)"
                jLogger.logError(log)
                product = products.last
            }
            
            finishCallback?(result: JResult.value(product!))
        } else {
        
            let invalidIdentifier = response!.invalidProductIdentifiers.last! as? String
        
            let errorClass = invalidIdentifier == productIdentifier
                ?JStoreKitInvalidProductIdentifierError.self
                :JStoreKitCanNoLoadProductError.self
        
            let error =  errorClass(productIdentifier: productIdentifier)
            finishCallback?(result: JResult.error(error))
        }
    }
    
    func request(request: SKRequest!, didFailWithError error: NSError!) {
        
        let passError = { () -> NSError in
            
            if let error = error {
                return error
            }
            
            return JSilentError(description:"SKRequest no inet connection")
        }
        finishCallback?(result: JResult.error(error))
    }
    
    @objc func productsRequest(request: SKProductsRequest!, didReceiveResponse response: SKProductsResponse!) {
        
        self.response = response
    }
}

//TODO JAsyncTypes<SKRequest>.JAsync
public func skProductLoaderWithProductIdentifier(identifier: String) -> JAsyncTypes<SKProduct>.JAsync
{
    let factory = { () -> JAsyncSKProductsRequestAdapter in
        return JAsyncSKProductsRequestAdapter(productIdentifier:identifier)
    }
    return JAsyncBuilder.buildWithAdapterFactory(factory)
}
